### ISSUE TITLE
No way to stop monitoring the socket

### ISSUE THREAD

**Issue by dshulepov** [zeromq/netmq#398]
If I understand correctly, then the following code will create a pair socket, bind it to "inproc://#monitor" endpoint and send all the events (which you asked to monitor) from the monitored socket to this pair socket.

```
socket.Monitor("inproc://#monitor", SocketEvents.All);
```

Using NetMQMonitor one could then connect to "inproc://#monitor" and "listen" to events on the `socket`.
However there is no way to stop monitoring the socket.
`SocketBase.Monitor` has a code to stop monitoring the socket if you pass a null endpoint, but the `NetMQSocket.Monitor` does not allow to pass null endpoint and throws and ArgumentNullException.
Also, since the disposing of sockets is async, calling `socket.Dispose()` does not free the "inproc://#monitor" endpoint and you can't reuse it to monitor the other socket.

The reason I found it (and may be I was doing something wrong) was because I was trying to do some sort of the reliability. I use a Request socket on the client for heartbeats. I tried to monitor it to get notified about disconnects and connection retries. However when you send a message with Request socket it switches to "receive state" and you can't send another message with it, so you should re-create the socket. This is where the problem with monitor hits you. You dispose the socket, but the endpoint for monitor remains occupied in the context and you can't monitor the new socket with the same endpoint.

**Comment by sharpe5** [comment]
+1 for this issue. I ran into it when creating NetMQ.ReactiveExtensions, as I needed to be sure that the socket was ready by listening to events from it.

My solution was to leave the monitor in place, but unsubscribe from the event handlers.

The best solution would be to allow the NetMQMonitor to be disposed without this inadvertently also disposing of its parent socket.

**Comment by somdoron** [comment]
I actually don't like monitoring (little sorry I port them :-)). Maybe we
can solve the problem in a better way. What is it that you solved with
monitoring?

**Comment by sharpe5** [comment]
Coincidentially, I just discussed this very topic. 

When I wrote NetMQ.ReactiveExtensions, I found that I needed to add a delay of 500 milliseconds after starting both the subscriber (and the publisher) to give them time to start up, or else they would sometimes lose the initial messages in the speed test. 

I also experimented with adding event handlers to wait for the socket to be ready, but I never managed to get this working reliably, and ended up adding the 500ms delay as well. It would be great if there was an easy way to wait the right time to be 100% sure that the NetMQ sockets were ready to go. 

It is interesting to note that all of the unit tests relating to pub sub have a 500ms delay. If this delay is removed, the unit tests wont pass. This delay worries me, as it could fail on a slower machine or if a Windows driver made everything pause for a bit. Thats why I use a combination of waiting for events, and a 500ms delay, to make sure that the socket is ready in NetMQ.ReacticeExtensions.
